File: Algorithm/220915/SWEA_사칙연산.py
import sys
sys.stdin = open('input.txt')

def postorder(cur):                         # 후위 순회로 검색
    if lst_l[cur] != 0:                     # 연산자라면 = 자식 노드가 있다면
        postorder(lst_l[cur])               # 먼저 좌, 우 노드를 모두 탐색 후...
        postorder(lst_r[cur])
        b = stk.pop()                       # 2개의 숫자를 스택에서 뽑음
        a = stk.pop()
        if lst_m[cur] == '+':               # 연산자에 따라 연산 처리 후 다시 스택에 저장
            stk.append(a + b)
        elif lst_m[cur] == '-':
            stk.append(a - b)
        elif lst_m[cur] == '*':
            stk.append(a * b)
        else:
            stk.append(a // b)
    else:                                   # 숫자라면
        stk.append(lst_m[cur])              # 그 위치의 숫자를 스택에 저장

for cs in range(10):
    n = int(input())
    lst_m, lst_l, lst_r = [0], [0], [0]
    stk = []
    for i in range(n):
        tup = input().split()
        if tup[1] in ['+', '-', '*', '/']:  # 연산자라면 입력 변수가 4개이므로
            lst_m.append(tup[1])            # 본인 노드, 좌우 자식 노드를 기록
            lst_l.append(int(tup[2]))
            lst_r.append(int(tup[3]))

        else:                               # 숫자라면 본인 노드를 기록 후, 자식 노드는 없다는 뜻으로 0을 기록
            lst_m.append(int(tup[1]))
            lst_l.append(0)
            lst_r.append(0)
    postorder(1)
    print(f'#{cs + 1} {stk[-1]}')           # 전부 완벽하게 수행됐다면 스택에는 최종 결과 숫자 하나만 들어있을 것

File: 220930_Algorithm/220915/SWEA_사칙연산.cpp
#include <iostream>
//#include <fstream>
#include <string> // 도~~~~~~저히 char 형으로는 안 되겠음. string으로 합니다.
using namespace std;

int i, n, node, lnode, rnode, cal_a, cal_b;
int lst_l[1001], lst_r[1001], stk[1001];
string dat, lst_m[1001];
int num_stk;

void postorder(int cur) {
	if (lst_l[cur] != 0) {
		postorder(lst_l[cur]);
		postorder(lst_r[cur]);
		num_stk--; cal_b = stk[num_stk]; // 스택 pop 연산
		num_stk--; cal_a = stk[num_stk];
		if (lst_m[cur] == "+") {
			stk[num_stk] = cal_a + cal_b; num_stk++;
		}
		else if (lst_m[cur] == "-") {
			stk[num_stk] = cal_a - cal_b; num_stk++;
		}
		else if (lst_m[cur] == "*") {
			stk[num_stk] = cal_a * cal_b; num_stk++;
		}
		else {
			stk[num_stk] = cal_a / cal_b; num_stk++;
		}
	}
	else {
		stk[num_stk] = stoi(lst_m[cur]); num_stk++;
	}
}
int main() {
	//ifstream cin("input.txt");
	for (int cs = 0; cs < 10; cs++) {
		cin >> n;
		for (i = 0; i < n; i++) {
			cin >> node >> dat;
			lst_m[node] = dat;
			if (dat == "+" || dat == "-" || dat == "*" || dat == "/") {
				cin >> lnode >> rnode;
				lst_l[node] = lnode;
				lst_r[node] = rnode;
			}
			else {
				lst_l[node] = 0;
				lst_r[node] = 0;
			}
		}
		num_stk = 0;
		postorder(1);
		cout << "#" << cs + 1 << " " << stk[0] << endl;
	}
	//cin.close();
	return 0;
}

// 
// 1. String 헤더 사용 + char to String 구문 익히기
// 2. char 관련 코드 새로 배우기 => 는 지금은 힘들듯

// 바보짓한 거
// 1. cout 출력은 어차피 처음부터 다 붙어서 나옴
// 2. char 배열은 생각보다 똑똑함. 기본적으로 null로 초기화돼서
// ...그냥 cout 해서 바로 출력해도 알아서 자료 넣은 곳까지 출력됨